contracts/nft_royalty/lib.rs

```
@@ -0,0 +1,437 @@
+//! - `create_transfer`: Adds a settlement instruction. The instruction will have three legs. One [`Leg`] where [`NFTTransferDetails::nft_owner_portfolio`] is transferring [`NFTTransferDetails::nfts`] to [`NFTTransferDetails::nft_receiver_portfolio`], another leg where [`NFTOffer::payer_portfolio`] sends [`NFTOffer::transfer_price`] to [`NFTOffer::receiver_portfolio`], and one leg where the payer is transferring the royalty to the artist.
+//! - `create_custody_portfolio`: Creates a portoflio owned by the contract's caller and transfer its custody to the smart contract.
+
+#![cfg_attr(not(feature = "std"), no_std, no_main)]
+
+extern crate alloc;
+
+use alloc::collections::BTreeSet;
+use alloc::string::{String, ToString};
+use alloc::vec::Vec;
+use ink::env::debug_println;
+use ink::storage::Mapping;
+use scale::{Decode, Encode};
+use sp_arithmetic::per_things::Perbill;
+
+use polymesh_ink::{
+    AssetMetadataKey, AssetMetadataLocalKey, AssetMetadataName, AssetMetadataValue, IdentityId,
+    Leg, NFTId, NFTs, PolymeshEnvironment, PolymeshError, PolymeshInk, PortfolioId, PortfolioName,
+    Ticker, VenueDetails, VenueId, VenueType,
+};
+
+pub use crate::nft_royalty::types::{NFTArtistRules, NFTOffer, NFTTransferDetails};
+
+#[ink::contract(env = PolymeshEnvironment)]
+mod nft_royalty {
+    use super::*;
+    use alloc::vec;
+
+    /// The asset metadata name for the key that holds the mandatory NFT collection metadata.
+    const NFT_METADATA_NAME: &[u8] = "v0_nft_madantory_metadata".as_bytes();
+    /// The details of the venue controlled by the smart contract.
+    const VENUE_DETAILS: &[u8] = "sc-controlled-venue".as_bytes();
+
+    /// The contract's result type.
+    pub type Result<T> = core::result::Result<T, Error>;
+
+    /// Contract Errors.
+    #[derive(Debug, Decode, Encode)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub enum Error {
+        /// Polymesh ink error.
+        PolymeshInk(PolymeshError),
```

We might be able to improve the errors from `polymesh_ink::PolymeshError`. It seems like this is hiding the error messages.

---

Yup.  I think previously it wasn't as easy to get the error value back.